apps/site/pages/en/blog/vulnerability/march-2025-ci-incident.md

```diff
@@ -1,13 +1,72 @@
+
+According to the HackerOne report, the exploit proceeded as follows:
+
+1. Submit a valid pull request to nodejs/node.
+2. Wait for a maintainer to add the `request-ci` label (this label is added to every pull request with non-documentation changes).
+3. After approval, update the pull request using an outdated Git commit timestamp.
+4. When Jenkins pipelines trigger, they fetch and execute code from the forked pull request.
+5. Attain code execution on Node.js Jenkins agents.
+
+Upon review, we identified that the `request-ci` label step simplifies but is **not required** to carry out the attack. A similar attack could be used against the `commit-queue` label, thus potentially allowing an attacker to land an unauthorized code change.
+
+The core issue stems from a Time-of-Check-Time-of-Use (TOCTOU) vulnerability between initiating a CI build and the moment the Jenkins job checks out the code. Previously, CI jobs used Git references (`refs/pull/<pr_id>/head`), which attackers can alter after triggering the CI. Importantly, the collaborator initiating the CI build did nothing wrong—the pull request appeared safe when CI was triggered.
+
+![Example of workflow for starting Jenkins CI on a Github Pull Request][example_test_infra]
+
+![Example of attack in the Node.js test infra][example_attack_test_Infra]
+
+## **Remediation**
+
+In response to this security incident, the Node.js security team took measures to mitigate risks and secure the infrastructure.
+
+- Immediately upon confirmation of the vulnerability, access to initiate new Jenkins CI runs was restricted to prevent further compromise while the team validated the report.
+- All compromised hosts (24 machines) were swiftly identified, removed from Jenkins, and rebuilt to eliminate any potential residual risk left over from the initial ingress.
+- Security improvements were implemented in Jenkins jobs to validate commit SHAs, ensuring jobs only executed trusted and verified code.
+- `request-ci` and `commit-queue` labels now act relying on validated commit SHAs instead of comparing dates.
+- Comprehensive audits were carried out across 140 Jenkins jobs, prioritizing frequently used ones, to detect and remediate vulnerabilities.
+- Identified vulnerable GitHub workflows were temporarily disabled, promptly patched, and re-enabled with enhanced security measures.
 
+These targeted actions significantly strengthened the security posture of our CI infrastructure, preventing the recurrence of similar potential
+intrusions and ensuring safe operations moving forward.
+
+## **Timeline**
+
+- **Friday, 21 March 2025**: Report received on Hackerone. Initial triage confirmed the report as a genuine issue. The ability to start new Jenkins CI runs was restricted to prevent any further machine compromises.
```

Comment:
According to the Microsoft Style Guide, we should be using numbered list for sequential items https://learn.microsoft.com/en-us/style-guide/scannable-content/lists#numbered-lists
```suggestion
1. **Friday, 21 March 2025**: Report received on Hackerone. Initial triage confirmed the report as a genuine issue. The ability to start new Jenkins CI runs was restricted to prevent any further machine compromises.
```